### PR TITLE
repro: check for hash mismatch between deleted dependencies and upstream outputs.

### LINKED ISSUE #9530: --allow-missing: skips stages even if deps have changed
# Bug Report

## Description

`--allow-missing` is skipping stages unexpectedly (in cases where pulling the data and then running the pipeline will not skip the stages).

### Reproduce

In https://github.com/dberenbaum/example-get-started-experiments/tree/split-0.2, I changed the param `data_split.test_pct` from `0.15` -> `0.2` and ran only the `data_split` stage.

```
$ git clone -b split-0.2 git@github.com:dberenbaum/example-get-started-experiments.git
$ cd example-get-started-experiments
$ dvc exp run --allow-missing --dry
Reproducing experiment 'faint-kine'
'data/pool_data.dvc' didn't change, skipping
Stage 'data_split' didn't change, skipping
Stage 'train' didn't change, skipping
Stage 'evaluate' didn't change, skipping
$ dvc pull
$ dvc exp run --dry
Reproducing experiment 'extra-colt'
'data/pool_data.dvc' didn't change, skipping
Stage 'data_split' didn't change, skipping
Running stage 'train':
> python src/train.py

Running stage 'evaluate':
> python src/evaluate.py
```

### Expected

I would expect that the `train` and `evaluate` stages would be run with or without `--allow-missing`, and that `--allow-missing` would run the same stages as it would if I first pulled and then ran without `--allow-missing`.

### Environment information

**Output of `dvc doctor`:**

```console
$ dvc doctor
DVC version: 2.58.1 (pip)
-------------------------
Platform: Python 3.10.10 on macOS-13.3.1-arm64-arm-64bit
Subprojects:
        dvc_data = 0.51.0
        dvc_objects = 0.22.0
        dvc_render = 0.5.3
        dvc_task = 0.2.1
        scmrepo = 1.0.3
Supports:
        azure (adlfs = 2023.4.0, knack = 0.10.1, azure-identity = 1.12.0),
        gdrive (pydrive2 = 1.15.3),
        gs (gcsfs = 2022.11.0),
        hdfs (fsspec = 2022.11.0, pyarrow = 11.0.0),
        http (aiohttp = 3.7.4.post0, aiohttp-retry = 2.8.3),
        https (aiohttp = 3.7.4.post0, aiohttp-retry = 2.8.3),
        oss (ossfs = 2021.8.0),
        s3 (s3fs = 2022.11.0, boto3 = 1.24.59),
        ssh (sshfs = 2023.4.1),
        webdav (webdav4 = 0.9.8),
        webdavs (webdav4 = 0.9.8),
        webhdfs (fsspec = 2022.11.0)
Config:
        Global: /Users/dave/Library/Application Support/dvc
        System: /Library/Application Support/dvc
Cache types: reflink, hardlink, symlink
Cache directory: apfs on /dev/disk3s1s1
Caches: local
Remotes: https
Workspace directory: apfs on /dev/disk3s1s1
Repo: dvc, git
Repo.site_cache_dir: /Library/Caches/dvc/repo/2ede911c0d219722caad26f44649bc87
```

### PR DESCRIPTION
**Pull request by daavoo**
Only if `--allow-missing` is passed.

- Create `tests/func/repro` and extract `pull` and `allow_missing` to separate test files.

Closes #9530

### CODE REVIEW

File: dvc/stage/__init__.py [hunk 1]
```diff
@@ -295,22 +295,33 @@ def env(self) -> Env:
...
         for dep in self.deps:
             status = dep.status()
             if status:
                 if allow_missing and status[str(dep)] == "deleted":
-                    continue
+                    if upstream and any(
+                        dep.fs_path == out.fs_path and dep.hash_info != out.hash_info
+                        for stage in upstream
+                        for out in stage.outs
+                    ):
```
Comment (daavoo) [inline dvc/stage/__init__.py:321]: This is not pretty but I don't see any alternative

**Comment by dberenbaum** [comment]
This works, so not a blocker, but is there a way we can actually have `dvc status --allow-missing` return `modified` here like discussed in the issue?

It sounds like something I would expect to be possible, especially with the index changes from @efiop that I thought should make it easy to compare against virtual outputs.

**Comment by daavoo** [comment]
> This works, so not a blocker, but is there a way we can actually have `dvc status --allow-missing` return `modified` here like discussed in the issue?

I think this P.R. should be merged and backported whereas the `status` change still needs discussion and would be a 3.0 change.

> especially with the index changes from @efiop that I thought should make it easy to compare against virtual outputs.

I am not sure I follow how the index changes have anything to do with this, as this just involves checking upstream stages in the pipeline

**Comment by dberenbaum** [comment]
> I think this P.R. should be merged and backported whereas the `status` change still needs discussion and would be a 3.0 change.

Okay, I'm fine to do it this way.

> I am not sure I follow how the index changes have anything to do with this, as this just involves checking upstream stages in the pipeline

I wasn't that familiar with how status works and thought it would be possible. In theory, we could build a virtual index based on the outputs of the `dvc.lock` and the `.dvc` files and do a status check against that rather than against the workspace. From a quick look, it looks like it would basically do the same as this method except not against the workspace (which I don't see an easy way to do):

https://github.com/iterative/dvc/blob/3e0f2bafbbafe4c66cc74ff116692052b4d49a03/dvc/output.py#L590-L600 

It would return a modified status when an upstream stage changed (since the hash of the out will change) without manually comparing upstream stages and changing the status inside repro.

**Comment by dberenbaum** [comment]
@daavoo Do you plan to merge as is? If so, can we get someone to take a look so we can unblock it please?

**Comment by daavoo** [comment]
> @daavoo Do you plan to merge as is? If so, can we get someone to take a look so we can unblock it please?

I did not plan to implement a different solution for the release if this one works. cc @iterative/dvc could you review this?

**Comment by skshetry** [comment]
I was trying to merge, but noticed that some tests were failing. I have rerun them, waiting for them to be green. :)

**Pull request merged**